### DISCUSSION TITLE
Reflection and CDI Issues with SmallRye GraphQL and Hibernate Integration

### DISCUSSION THREAD

**Discussion by x80486** [quarkusio/quarkus#43102] [Q&A]
I'm experiencing some issues related to reflection and CDI when using SmallRye GraphQL and Hibernate in a Quarkus application.

I think these types should be managed automatically by Quarkus, as they are part of the supported integrations, but I'm not entirely sure.

This is the output when running the application:

```shell
Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
        - io.smallrye.graphql.client.typesafe.api.ErrorOr (source: <unknown>)
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.

================================================================================
CDI: programmatic lookup problem detected
-----------------------------------------
At least one bean matched the required type and qualifiers but was marked as unused and removed during build

Stack frame: io.quarkus.hibernate.orm.runtime.customized.FormatMapperKind$2.create(FormatMapperKind.java:27)
Required type: interface jakarta.json.bind.Jsonb
Required qualifiers: [@jakarta.enterprise.inject.Default()]
Removed beans:
        - PRODUCER_METHOD bean io.quarkus.jsonb.JsonbProducer#jsonb() [types=[interface java.lang.AutoCloseable, interface jakarta.json.bind.Jsonb], qualifiers=null]
Solutions:
        - Application developers can eliminate false positives via the @Unremovable annotation
        - Extensions can eliminate false positives via build items, e.g. using the UnremovableBeanBuildItem
        - See also https://quarkus.io/guides/cdi-reference#remove_unused_beans
        - Enable the DEBUG log level to see the full stack trace
================================================================================
```

I believe this error is somehow coming because this piece of code:

```java
import jakarta.inject.Inject;
import org.eclipse.microprofile.graphql.GraphQLApi;
import org.eclipse.microprofile.graphql.GraphQLException;
...

@GraphQLApi
public class GQLAPI {
  ...

  @Query("get")
  public MyEntity get(@Name("id") final UUID id) throws GraphQLException { // Bonus for making GraphQLException unchecked! :partying_face:
    return service.getByID(id)
        .orElseThrow(() -> new GraphQLException("unable to retrieve record));
  }
}
```

This is the only place where `io.smallrye.graphql.client.typesafe.api.ErrorOr` would be used (internally), and because it's a built-in type from SmallRye GraphQL, I would think it should be managed by Quarkus, rather than creating entries in `application.yaml` for the same.

[a comment by an automated account is not shown]

**Comment by gsmet** [comment]
I think there are two different issues here: one in the SmallRye GraphQL extension and one in the Hibernate ORM extension.

Any chance you could provide a small Maven app that reproduces the problems?

**Reply by gsmet** [reply]
I created https://github.com/quarkusio/quarkus/issues/43105 for the unremovable bean part. Not sure what's going on with SmallRye GraphQL so reproducer welcome.

**Reply by x80486** [reply]
@gsmet, thanks for looking into it.

[quarkus-smallrye-graphql-hibernate-orm.zip](https://github.com/user-attachments/files/16919288/quarkus-smallrye-graphql-hibernate-orm.zip)  is a small project where you can see this behavior.

I use `./mvnw clean verify` and `./mvnw quarkus:dev` — both show the same issue/warning, but the tests provide _some more_ context.

**Comment by SteveGueldner** [comment]
Same problem here. Just to isolate it a little more I removed all JsonB code from my project. The problem persists. It seems having JsonB in the dependencies is enough to trigger this behavior.
